providers/salesforce/bulk-info.go

```diff
@@ -7,8 +7,88 @@ import (
 	"net/http"
 
 	"github.com/amp-labs/connectors/common"
+	"github.com/amp-labs/connectors/common/handy"
 )
 
+type ListIngestJobsResult struct {
+	Done           bool               `json:"done"`
+	NextRecordsURL string             `json:"nextRecordsUrl"`
+	Records        []GetJobInfoResult `json:"records"`
+}
+
+// ListIngestJobsInfo returns information about Ingest Jobs. If jobIds are provided, only those jobs are returned.
+// It is possible to get information about all current ingest jobs by not providing any jobIds. Note that Salesforce
+// returns terminal state jobs from a maximum of 7 days ago.
+// nolint:funlen,cyclop
+func (c *Connector) ListIngestJobsInfo(ctx context.Context, jobIds ...string) ([]GetJobInfoResult, error) {
+	url, err := c.getRestApiURL("jobs/ingest")
+	if err != nil {
+		return nil, err
+	}
+
+	// Filters out BigObjects / BulkAPI v1 jobs
+	url.WithQueryParam("jobType", "V2Ingest")
+
+	// To collect all the jobs we need to return
+	var jobsInfo []GetJobInfoResult
+
+	// If we have jobIds, we create a set to keep track of the matches we need to find. Each time we get
+	// a match, we remove it from the set. If the set is empty, we can break the loop to save time and unnecessary
+	// pagination.
+	pending := handy.NewSet[string](jobIds)
+
+	// To keep track of pages
+	location := url.String()
+
+	for {
+		res, err := c.Client.Get(ctx, location)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to list ingest jobs info: %w",
+				errors.Join(err, common.ErrRequestFailed),
+			)
+		}
+
+		response, err := common.UnmarshalJSON[ListIngestJobsResult](res)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to unmarshal list ingest jobs info: %w",
+				errors.Join(err, common.ErrParseError),
+			)
+		}
+
+		// Add the jobs we need to the list (or all of them if we don't have a filter)
+		for _, result := range response.Records {
+			if len(jobIds) == 0 || contains[string](jobIds, result.Id) {
+				jobsInfo = append(jobsInfo, result)
+
+				// This is a no-op if we don't have jobIds, or if the job is not in the set.
+				pending.Remove(result.Id)
+			}
+		}
+
+		if response.Done {
+			break
+		}
+
+		// If we aren't done yet, check if we have all the jobs we need
+		if len(jobIds) > 0 && pending.IsEmpty() {
+			break
+		}
+
+		// Else, get the next page
+		domain, err := c.getDomainURL()
+		if err != nil {
+			return nil, err
+		}
+
+		// getDomainURL escapes some of the characters in the nextRecordsURL, so we use the URL directly
+		location = domain.String() + response.NextRecordsURL
```

Comment:
I am not sure what is the content of `response.NextRecordsURL`.
There is `domain.AddPath(...strings...)` which uses `url.URL.JoinPaith` underneath.
Maybe it can be useful here. If it is then we need would need to keep `getDomainURL` inside loop as we would create a new URL instance and modify it.

Reply:
I tried to use that initially, but it kept escaping ? to a %3F

Reply:
Merging this now, if there's any helper that accomplishes the same, we can change it later

Reply:
Yeah, I also noticed that `url.URL.JoinPaith` doesn't have that many usecases. For resolving this super strict escaping I added `AddEncodingExceptions(...)` but I would rather stick to string concat. I wonder why standard `url.Url` doesnt deal with it.